debian/patches/0007-add-bt2390-eetf-and-code-refactor-to-opencl-tonemap.patch

```diff
@@ -1644,7 +1652,36 @@ Index: FFmpeg/libavfilter/vf_tonemap_opencl.c
      if (ctx->range_in == AVCOL_RANGE_JPEG)
          av_bprintf(&header, "#define FULL_RANGE_IN\n");
  
-@@ -199,19 +538,41 @@ static int tonemap_opencl_init(AVFilterC
+     if (ctx->range_out == AVCOL_RANGE_JPEG)
+         av_bprintf(&header, "#define FULL_RANGE_OUT\n");
+ 
++    if (ctx->in_desc->comp[0].depth == 10) {
++        av_bprintf(&header, "#define INPUT_QUANTIZATION_OFFSET 0.0009765774014f\n");
++        av_bprintf(&header, "#define INPUT_Y_SCALE 1.1678082192f\n");
++        av_bprintf(&header, "#define INPUT_UV_SCALE 1.1417410714f\n");
++    } else if (ctx->in_desc->comp[0].depth == 16) {
++        // 16bit texture is actually 12bit
```

Comment:
Actually, for the method of using P016 to store 12bit images produced by HW decoder, you cannot use pixdesc to distinguish the actual valid bits. For example, `-vf format=p016,hwupload,tonemap_gpu...`

It’s just that the current `scale_{cuda,opencl}` does not support yuv2rgb and vice versa, so this issue has not been raised in the upstream.

Reply:
I also have some concerns about vendor-specific behaviors in the texture normalization, like will they automatically shift the signal so that all of them fully utilizing the 16bit range, but probably they don't.

Reply:
On these GPGPU platforms for numerical computing, the driver should only compute the given data according to the depth/format you specify, even if the data is mismatched they will be computed as is. It's unlikely that opaque operations like the one you mentioned will occur.

Reply:
I choose to assume the input is still 12bit because:
- It is what hardware decoders produce
- Videos are allowed up to 12 bit for now in BT.2100 standard
- We cannot tell the actual video quantization depth if it is altered during the pipeline like using `format=p016`, so the color will always be off.